File: crm_ui_client/src/api/crm/dto/Application.ts
export enum ApplicationStatus {
    Pending, Accepted, Aborted
}

export interface ApplicationRawData {
    professionalId: number
    jobOfferHistoryId: number
    status: ApplicationStatus
    date: string
}

export class Application implements ApplicationRawData {
    professionalId: number
    jobOfferHistoryId: number
    status: ApplicationStatus
    date: string

    constructor(professionalId: number, jobOfferHistoryId: number, status: ApplicationStatus, date: string) {
        this.professionalId = professionalId
        this.jobOfferHistoryId = jobOfferHistoryId
        this.status = status
        this.date = date
    }

    static fromJsonObject(obj: ApplicationRawData): Application | null {
        try {
            return new Application(obj.professionalId, obj.jobOfferHistoryId, obj.status, obj.date)
        } catch (e) {
            console.error(e)
            return null
        }
    }
}

File: crm/src/main/kotlin/it/polito/crm/entities/Application.kt
package it.polito.crm.entities

import it.polito.crm.dtos.ApplicationDTO
import jakarta.persistence.*
import java.io.Serializable
import java.time.LocalDateTime

enum class ApplicationStatus {
    Pending, Accepted, Aborted
}

class ApplicationId(
    var professional: Long = 0,
    var jobOfferHistory: Long = 0
) : Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        other as ApplicationId

        if (jobOfferHistory != other.jobOfferHistory) return false
        if (professional != other.professional) return false

        return true
    }

    override fun hashCode(): Int {
        var result = jobOfferHistory.hashCode()
        result = 31 * result + professional.hashCode()
        return result
    }
}

@Entity
@Table(name = "applications")
@IdClass(ApplicationId::class)
class Application {
    @Id
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "professional_id", referencedColumnName = "professional_id")
    lateinit var professional: Professional

    @Id
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "job_offer_history_id", referencedColumnName = "job_offer_history_id")
    lateinit var jobOfferHistory: JobOfferHistory

    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    var status: ApplicationStatus = ApplicationStatus.Pending

    @Column(name = "date")
    val date: LocalDateTime = LocalDateTime.now()

    fun toDto(): ApplicationDTO {
        return ApplicationDTO(
            this.professional.professionalId,
            this.jobOfferHistory.jobOfferHistoryId,
            this.status,
            this.date
        )
    }
}